### DISCUSSION TITLE
Search results for BitClout.com

### DISCUSSION THREAD

**Discussion by daniel-keyes** [deso-protocol/dips#58] [Medium CIPs]
Hi 👋

Search is essential to navigating every modern social network. It is critical in helping us find and interact with other people, topics of interest, and specific posts. Today, BitClout.com does not have search or discovery built-in. 

Cloutavista would like to integrate our search capabilities directly within BitClout.com.
We've built an API and can return a results page in the same format as the current Feed page. 

**Search bar:**
![big bar](https://user-images.githubusercontent.com/87264202/125591468-a8ec09ec-9175-4fc2-87b2-bc234001f130.png)

**Results page with "Top" and "Recent" feed options:**
![Group 2471](https://user-images.githubusercontent.com/87264202/125591692-ee1ac3f8-6e29-449d-8563-f999d0f2d86b.png)

**Scrolled results page:**
![Group 248](https://user-images.githubusercontent.com/87264202/125591776-247b95be-98b6-45a8-a013-22e455556fb3.png)

Would love to hear your thoughts on what features to add/ the design. 

Thanks!

**Comment by ItsAditya-xyz** [comment]
I like it!
So are we going to have search filters as well? Just like we have in Cloutavista where I can filter my search just like how the search works in discord ( from, has, when)

**Reply by daniel-keyes** [reply]
This is definitely doable, but it might be best to start with a simple search and redirect for advanced filters.

**Comment by tijno** [comment]
Search is important so I am up for it getting added.

The mockups seems to cover anything i would expect.

I dont think it needs additional filters - people can use cloutavista for that if they want. 

But I think your discussion needs to have more detail about what you are proposing and the implications of it. The CIP template is a good starting point.

In addition, would be good to get clarity on the following:

Eg:
* Will you track users as you do on cloutavista?
* Are there any commercial implications - eg if nodes want to use the same search feature, will they have to buy your services? If so what are the terms around this?
* Does it surface posts that are hidden?

Also re implementation 

* Maybe there is an opportunity to add a generic search request api that allows nodes and users to select their preferred search engine.

* I think the api should just return the postHash and not the full posts. The frontend will need to make a call to its own backend anyway to check for hidden posts or blocked users so they can filter those out.

* Currently there is no api endpoint in backend to get post data for an array of posts, but I think this could be easily added to the get posts stateless. The logic already exists in get user stateless.

**Reply by iPaulPro** [reply]
It's a non-starter for frontend to directly hit a third party API. Even the way we use https://pulse.bitclout.com/api/, instead of hitting Pulse directly, is a stretch. We just trust these third parties not to game these lists, and to always be available? 

That's not how this decentralized thing is supposed to work.

**Comment by daniel-keyes** [comment]
@AdityaChaudhary0005 I think I agree with @tijno that at least initially we can reroute advanced filter searches. I think it would keep the functionality simpler and have a more seamless UX/UI.

@tijno, you bring up some really good points and questions. Re the implementation, I agree that the easiest way would be to return a list of postHashes. 

I see two options for tracking. The simplest and perhaps best would be to have no user tracking through search. If however, the community wanted personalized results (for example, seeing posts from followers, coin holders, etc) we could add the public key as part of the search query to Cloutavista. 

We are also working with other large nodes to give them access to our API for search and discovery. It's still early (like everything on BitClout), so we are not entirely sure how or even if we will commercialize.

**Reply by tijno** [reply]
Tracking should be disabled. 
Post hashes is the way to go - this will overcome issue of stake post data or isHidden setting.
Details on commercial terms or clear commitment to not charge nodes for the service would be good.

**Reply by daniel-keyes** [reply]
Yes, I agree 100%. There will be no tracking at all. 
More so, if we only supply the Post hashes, there is no vulnerability to security issues.

**Comment by daniel-keyes** [comment]
@lazynina @maebeam @diamondhands0 any thoughts on a native search results integration? We can have NFT search available for launch as well.

**Reply by tijno** [reply]
I think it would be good to see more detail about how you propose to implement the change?

**Comment by iPaulPro** [comment]
Wait, so we're talking about integrating a closed-source third-party search solution, or will Cloutavista be open sourced and properly integrated into frontend?

Integrating closed-source APIs seems to firmly go against the ideals of Bitclout; open source and decentralized. Why would we want to rely on a random party for a service like this, leaving all other nodes to roll their own solution or be beholden to this random party?

I vote for a solution that is fully open and part of frontend.

**Reply by iPaulPro** [reply]
Here's a great example of the dangers of using a closed-source solution: 

<img width="500" src="https://user-images.githubusercontent.com/354227/126777091-c80aa750-c4ab-4190-9d09-9e586d4cee9a.jpg">

https://bitclout.com/posts/ac43defe37f6e78beca7c7834db0f7bd95a82791a2cdba7c281565e8f2143776

**Reply by daniel-keyes** [reply]
Huge thanks to @AdityaChaudhary0005 and the entire community for flagging the bug we had. Thanks to you, it was fixed in a matter of minutes.

@iPaulPro you bring up some really important points. I think there's a critical question here about how we, as a community, can build on top of BitClout so it continues to flourish and grow. 

My main goal is to build search and discovery tools that help the BitClout community and is sustainable over the long term.  
We plan on open-sourcing as much as possible and look forward to building in tandem with the community. We plan on raising money and hiring a team that can work on these tough problems full-time, and for that, we need to keep some of the aspects closed-source.

A major benefit of building social apps on top of an open protocol is that each node does not need to rebuild every component by itself, but rather can "outsource" them and integrate it with an API. We are already working with other Nodes and have supplied them with an API to power their search and discovery. Furthermore, I have seen other closed-source projects integrate with BitClout.com such as Pulse and BitHunt.

We welcome all of your ideas of how to make Cloutavista -- and all other BitClout projects -- sustainable over the long term.
Thank you all so much for the comments and ideas you have raised so far.
